boxoffice/models/line_item.py

```diff
@@ -73,6 +75,39 @@ def calculate(cls, line_item_dicts, coupons=[]):
             line_items.extend(item_line_items[item_id])
         return line_items
 
+    def confirm(self):
+        self.status = LINE_ITEM_STATUS.CONFIRMED
+
+
+def get_from_item(cls, item, qty, coupon_codes=[]):
+    """
+    Returns a list of (discount_policy, discount_coupon) tuples
+    applicable for an item, given the quantity of line items and coupons if any.
+    """
+    automatic_discounts = item.discount_policies.filter(DiscountPolicy.discount_type == DISCOUNT_TYPE.AUTOMATIC,
+        DiscountPolicy.item_quantity_min <= qty).all()
+    policies = [(discount, None) for discount in automatic_discounts]
+    if not coupon_codes:
+        return policies
+    else:
+        coupon_policies = item.discount_policies.filter(DiscountPolicy.discount_type == DISCOUNT_TYPE.COUPON).all()
+        coupon_policy_ids = [cp.id for cp in coupon_policies]
+        for coupon_code in coupon_codes:
+            coupon = DiscountCoupon.query.filter(DiscountCoupon.discount_policy_id.in_(coupon_policy_ids), DiscountCoupon.code == coupon_code).one_or_none()
+            if coupon and coupon.usage_limit > coupon.used_count:
+                policies.append((coupon.discount_policy, coupon))
+
+    return policies
+
+DiscountPolicy.get_from_item = classmethod(get_from_item)
+
+
+def update_used_count(self):
+    self.used_count = select([func.count(LineItem.discount_coupon_id)]).where(LineItem.discount_coupon_id == self.id).where(LineItem.status == LINE_ITEM_STATUS.CONFIRMED).as_scalar()
```

Comment:
I'd prefer to not use `id` columns for comparison so that if there's a type mismatch in your code, SQLAlchemy will catch it for you. In this case, this should be:

`self.used_count = select([func.count()]).where(LineItem.discount_coupon == self).where(LineItem.status == LINE_ITEM_STATUS.CONFIRMED).as_scalar()`